Separate face matching from _compute_facial_adjacency_from_vertices

majosm opened the pull request:
Pulled out of #204 to declutter.

:warning: Remove `_FlatFacialAdjacencyData` again before merging. :warning:

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1048,41 +1048,68 @@ def _boundary_tag_bit(boundary_tags, btag_to_index, boundary_tag):
-    is a :class:`numpy.ndarray` containing data for each stored face and its
-    adjacent neighbor.
+    Data structure for storage of a list of face identifiers (group, element, face).
+    Each attribute is a :class:`numpy.ndarray`.
```
Document shape of arrays.

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1048,41 +1048,68 @@ def _boundary_tag_bit(boundary_tags, btag_to_index, boundary_tag):
-        The group containing the adjacent element, or -1 if the face is not shared.
 
-    .. attribute:: neighbors
+def _match_faces_by_vertices(groups, face_ids, vertex_index_map_func=None):
```
Document return value.

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1091,56 +1118,46 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
-            face_vertex_indices[:len(ref_fvi), istart:iend] = grp_fvi.T
-            face_ids[0, istart:iend] = igrp
-            face_ids[1, istart:iend] = fid
+    matches = _match_faces_by_vertices(groups, face_ids)
```
`matches`: better variable name

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1091,56 +1118,46 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
+    matches_both_ways = np.stack((
+        np.concatenate((matches[0, :], matches[1, :])),
+        np.concatenate((matches[1, :], matches[0, :]))))
+    order = np.lexsort((matches_both_ways[1, :], matches_both_ways[0, :]))
+    sorted_matches = matches_both_ways[:, order]
```
Explain reasoning for sort.

inducer reviewed (commented):
Thanks! Some review bits that came up during our meeting.

inducer commented:
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

majosm commented:
Ready for a look @inducer.

I temporarily restored `_FlatFacialAdjacencyData` to make the diff a little less nasty. We'll have to remove it again before merging.

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
+class _FaceIDs:
+    """
+    Data structure for storage of a list of face identifiers (group, element, face).
+    Each attribute is a :class:`numpy.ndarray` of shape (nfaces,).
```
```suggestion
    Each attribute is a :class:`numpy.ndarray` of shape ``(nfaces,)``.
```

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
+def _match_faces_by_vertices(groups, face_ids, vertex_index_map_func=None):
+    """
+    Return matching faces in *face_ids*, where two faces match if they have the
+    same vertices. Expressed as a :class:`np.ndarray` of shape (2,nmatches) of
```
```suggestion
    same vertices. Expressed as a :class:`np.ndarray` of shape ``(2, nmatches)`` of
```

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
+
+def _match_faces_by_vertices(groups, face_ids, vertex_index_map_func=None):
+    """
+    Return matching faces in *face_ids*, where two faces match if they have the
```
> Return matching faces

In what form?

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
+    """
+    Return matching faces in *face_ids*, where two faces match if they have the
+    same vertices. Expressed as a :class:`np.ndarray` of shape (2,nmatches) of
+    indices into *face_ids*. *vertex_index_map_func* is used to map vertices to
+    other vertices; it must accept (possibly multidimensional) numpy arrays of
+    vertex indices.
```
Document arguments. Specifically, document the signature and return type (also acceptable shapes) of `vertex_index_map_func`.

Also document the return type and meaning of returned data from `_match_faces_by_vertices`.

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
+
+    # Lexicographically sort the face vertex indices, then diff the result to find
+    # faces with the same vertices
+    face_vertex_indices_increasing = np.sort(face_vertex_indices, axis=0)
```
```suggestion
    
    # normalize vertex-based "face identifiers" by sorting
    face_vertex_indices_increasing = np.sort(face_vertex_indices, axis=0)
```

inducer commented on meshmode/mesh/__init__.py:
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
+        np.concatenate((
+            face_index_pairs[1, :],
+            face_index_pairs[0, :]))))
+    # Sort by group, then neighbor group. Sort face indices because face_ids is
+    # already ordered by group
```
```suggestion
    # Accomplish a sort by group, then neighbor group. This is done by sorting by the indices
    # in face_ids. Realize that those are already ordered by group by construction.
```

inducer reviewed (commented):
Thanks! This looks great functionally. Just some nitpicking around docs and explanation, so that (hopefully) reading this becomes easier, then this is ready to go.

majosm commented:
> > Remove `_FlatFacialAdjacencyData` again before merging.
> 
> Don't see that anywhere? What's this referring to?

It's right above `_FaceIDs` in `meshmode/mesh/__init__.py`. I removed it because it isn't needed anymore, but then I added it back in fe50f623b4082 because removing it made the diff hard to follow.

inducer reviewed (approved):
Thanks! This LGTM now. I'll wait for you to remove `_FlatFacialAdjacencyData`. Let me know if/when it's ready to merge from your end.

majosm commented:
Good to go @inducer.

inducer commented:
Thanks! (Hope you're OK with me squashing this... if not, please tell me, although I won't be able to do much to fix it.)
